frame/treasury/src/lib.rs

```diff
@@ -531,9 +701,94 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		r
 	}
 
-	/// Spend some money! returns number of approvals before spend.
+	/// Spend_funds is triggered periodically and uses the `T::Paymaster` to payout all spend
+	/// requests in the `PendingPayments` storage map.
 	pub fn spend_funds() -> Weight {
 		let mut total_weight = Weight::zero();
+		let mut total_spent = BalanceOf::<T, I>::zero();
+		let mut missed_proposals: u32 = 0;
+		let proposals_len = PendingPayments::<T, I>::count();
+
+		Self::deposit_event(Event::ProcessingProposals { waiting_proposals: proposals_len });
+
+		for key in PendingPayments::<T, I>::iter_keys() {
+			if let Some(mut p) = PendingPayments::<T, I>::get(key) {
+				match p.payment_id {
+					None => match T::Paymaster::pay(&p.beneficiary, p.asset_kind, p.value) {
+						Ok(id) => {
+							total_spent += p.normalized_value;
+							p.payment_id = Some(id);
+							Self::deposit_event(Event::PaymentTriggered {
+								pending_payment_index: key,
+								asset_kind: p.asset_kind,
+								amount: p.value,
+								payment_id: id,
+							});
+							PendingPayments::<T, I>::set(key, Some(p));
+						},
+						Err(err) => {
+							log::debug!(target: LOG_TARGET, "Paymaster::pay failed for PendingPayment with index: {:?} and error: {:?}", key, err);
+							missed_proposals = missed_proposals.saturating_add(1);
+							Self::deposit_event(Event::PaymentFailure {
+								pending_payment_index: key,
+								asset_kind: p.asset_kind,
+								amount: p.value,
+								payment_id: None,
+							});
+						},
+					},
+					Some(payment_id) => match T::Paymaster::check_payment(payment_id) {
+						PaymentStatus::Failure => {
+							log::debug!(
+								target: LOG_TARGET,
+								"Paymaster::pay failed for PendingPayment with index: {:?}",
+								key
+							);
+							// try again in the next `T::SpendPeriod`.
+							missed_proposals = missed_proposals.saturating_add(1);
+							Self::deposit_event(Event::PaymentFailure {
+								pending_payment_index: key,
+								asset_kind: p.asset_kind,
+								amount: p.value,
+								payment_id: Some(payment_id),
+							});
+							// Force the payment to none, so a fresh payment is sent during the next
+							// T::SpendPeriod.
+							p.payment_id = None;
+							PendingPayments::<T, I>::set(key, Some(p));
+						},
+						PaymentStatus::Success => {
+							PendingPayments::<T, I>::remove(key);
+							Self::deposit_event(Event::PaymentSuccess {
+								pending_payment_index: key,
+								asset_kind: p.asset_kind,
+								amount: p.value,
+								payment_id,
+							});
+						},
+						// PaymentStatus::InProgress and PaymentStatus::Unknown indicate that the
+						// proposal status is inconclusive, and might still be successful or failed
+						// in the future.
+						PaymentStatus::InProgress | PaymentStatus::Unknown => {},
```

Comment:
if `Unknown` it will be never successful

Reply:
Are you sure about this? I'm hesitant to treat Unknown as a failure if there is a chance that the status could change at a subsequent point.